src/common/threadpool.cpp

```diff
@@ -0,0 +1,169 @@
+#include <voxen/common/threadpool.hpp>
+
+#include <thread>
+#include <atomic>
+#include <condition_variable>
+#include <future>
+#include <cassert>
+
+#include <voxen/util/log.hpp>
+
+namespace impl
+{
+
+struct ReportableWorkerState
+{
+	std::mutex semaphore_mutex;
+	std::condition_variable semaphore;
+	std::atomic_bool is_exit;
+
+	std::mutex state_mutex;
+	std::queue<std::packaged_task<void()>> tasks_queue;
+};
+
+struct ReportableWorker
+{
+	std::thread worker;
+	ReportableWorkerState state;
+};
+
+}
+
+namespace voxen
+{
+
+ThreadPool* ThreadPool::global_voxen_pool = nullptr;
+
+ThreadPool::ThreadPool(int start_thread_count)
+{
+	if (start_thread_count == -1) {
+		size_t std_hint = std::thread::hardware_concurrency();
+		if (std_hint == (size_t)0)
+			start_thread_count = DEFAULT_START_THREAD_COUNT;
+		else
+			start_thread_count = ((int)std_hint - 2); //2 already used by voxen: World thread and GUI thread
+	}
+
+	for (int i = 0; i < start_thread_count; i++)
+		run_worker(make_worker());
+}
+
+ThreadPool::~ThreadPool() noexcept
+{
+	for (impl::ReportableWorker* worker : m_workers)
+	{
+		worker->state.is_exit.store(true);
+		worker->state.semaphore.notify_one();
+	}
+
+	for (impl::ReportableWorker* worker : m_workers)
+		if (worker->worker.joinable())
+			worker->worker.join();
+
+	while (m_workers.size() > 0)
+	{
+		delete m_workers.back();
+		m_workers.pop_back();
+	}
+}
+
+void ThreadPool::workerFunction(impl::ReportableWorkerState* state)
+{
+	while(!state->is_exit.load())
+	{
+		{
+		std::unique_lock<std::mutex> lock(state->semaphore_mutex);
+		state->semaphore.wait(lock);
+		}
+
+		state->state_mutex.lock();
+		while (state->tasks_queue.size() != 0)
+		{
+			std::packaged_task<void()> task = std::move(state->tasks_queue.front());
+			state->tasks_queue.pop();
+			state->state_mutex.unlock();
+
+			task();
+		}
+		state->state_mutex.unlock();
+	}
+}
+
+impl::ReportableWorker* ThreadPool::make_worker()
+{
+	impl::ReportableWorker* new_worker = new impl::ReportableWorker();
+	new_worker->state.is_exit.store(false);
+	m_workers.push_back(new_worker);
+	return new_worker;
+}
+
+void ThreadPool::run_worker(impl::ReportableWorker* worker)
+{
+	std::packaged_task<void(impl::ReportableWorkerState*)> main_work_task(&ThreadPool::workerFunction);
+	worker->worker = std::move(std::thread(std::move(main_work_task), &worker->state));
```

Comment:
Thread construction can be simplified to `std::thread(&ThreadPool::workerFunction, &worker->state)`